velox/exec/WindowPartition.cpp

```diff
@@ -407,26 +407,63 @@ void WindowPartition::updateKRangeFrameBounds(
             mappedFrameColumn,
             flags) == 0) {
       rawFrameBounds[i] = rawPeerBounds[i];
-    } else {
-      // If the search is for a preceding bound then rows between
-      // [0, currentRow] are examined. For following bounds, rows between
-      // [currentRow, numRows()) are checked.
-      if (isPreceding) {
+      continue;
+    }
+
+    auto compareWithLastRow = [&](vector_size_t currentRow) -> int32_t {
+      if (currentRow == startRow) {
+        // For the first row, just returns an arbitrary positive value
+        // to indicate not to use the sliding optimization.
+        return 1;
+      }
+      return data_->compare(
+          partition_[currentRow - 1],
+          partition_[currentRow],
+          mappedFrameColumn,
+          mappedFrameColumn,
+          flags);
+    };
+
+    // If the search is for a preceding bound then rows between
+    // [0, currentRow] are examined. For following bounds, rows between
+    // [currentRow, numRows()) are checked. If the sliding search pattern
+    // is applicable, we can use the last row's search result to determine
+    // the start search index for current row.
+
+    auto compareResult = compareWithLastRow(currentRow);
```

Comment:
This code is tested by the cases here: https://github.com/facebookincubator/velox/blob/1bf58f4d2b2b5bd93f9c0a6690c57c58a574b0a2/velox/functions/lib/window/tests/WindowTestBase.cpp#L312

And these are all sliding window range frames. 

Could we add more tests for arbitrary frames now as your code needs to catch those ?